core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java

```diff
@@ -456,4 +473,61 @@ public void close() {
       // Nothing to clean up
     }
   }
+
+  /**
+   * Periodically check if we need to take daily master metadata snapshot.
+   */
+  private final class DailyBackupHeartbeatExecutor implements HeartbeatExecutor {
+    private final Pattern mBackupPattern
+        = Pattern.compile("alluxio-backup-([0-9]+-[0-9]+-[0-9]+)-[^.]+.gz");
+    private final SimpleDateFormat mDateFormat = new SimpleDateFormat("yyyy-MM-dd");
+    private Calendar mCalendar = Calendar.getInstance();
+
+    @Override
+    public void heartbeat() {
+      mCalendar.setTime(new Date());
+      int hour = mCalendar.get(Calendar.HOUR_OF_DAY);
+      if (hour != Configuration.getInt(PropertyKey.MASTER_DAILY_BACKUP_HOUR)) {
+        return;
+      }
+      String dir = Configuration.get(PropertyKey.MASTER_BACKUP_DIRECTORY);
+      BackupOptions options = new BackupOptions(dir, false);
+      try {
+        BackupResponse resp = backup(options);
+        AlluxioURI backupUri = resp.getBackupUri();
+        LOG.info("Successfully backed up journal to %s%n", backupUri);
+      } catch (Exception e) {
+        LOG.error("Failed to execute daily backup at {}", dir, e);
+      }
+      try {
+        deleteOutdatedBackups(new AlluxioURI(dir));
+      } catch (Exception e) {
+        LOG.error("Failed to delete outdated backup files", e);
+      }
+    }
+
+    /**
+     * Deletes outdated backup files to avoid consuming too many spaces.
+     *
+     * @param uri the backup directory uri
+     */
+    private void deleteOutdatedBackups(AlluxioURI uri) throws Exception {
```

Comment:
can we check how many backups are actually there, and only delete the oldest ones? e.g. if they shut down their cluster for 4 days, then start it up, we don't want to delete all backups.

Reply:
I remain 3 newest backups and delete the oldest ones. Should I increase the value? is 3 enough?
Thanks

Reply:
Can we make this configurable? I think 3 is a reasonable default